File: Advanced/PageAndRDL/OracleDataProvider/VB.NET/OracleDataProvider/OracleConnectionAdapter.vb
Imports GrapeCity.BI.Data.DataProviders
Imports Oracle.ManagedDataAccess.Client
Imports System.Data.Common
Imports System.Globalization

Namespace GrapeCity.ActiveReports.Samples.OracleDataProvider
	Public NotInheritable Class OracleConnectionAdapter
		Inherits DbConnectionAdapter

		Public Shared Instance As OracleConnectionAdapter = New OracleConnectionAdapter()

		Protected Overrides Function MultivalueParameterValueToString(ByVal parameterArrayValue As Object()) As String
			Return String.Join(",", parameterArrayValue.[Select](Function(parameterValue) "'" & Convert.ToString(parameterValue, CultureInfo.InvariantCulture) & "'"))
		End Function

		Protected Overrides Function CreateParameterNamePattern(ByVal name As String) As String
			If Not name.StartsWith(":") Then name = String.Format(":{0}", name)
			Return MyBase.CreateParameterNamePattern(name)
		End Function

		Public Overrides Sub SetParameters(queryParameters As IEnumerable(Of DbCommandParameter), command As DbCommand)
			If TypeOf command Is OracleCommand Then
				Dim oracleCommand As OracleCommand = command
				oracleCommand.BindByName = True
			End If

			MyBase.SetParameters(queryParameters, command)
		End Sub
	End Class
End Namespace

File: Advanced/PageAndRDL/OracleDataProvider/C#/OracleDataProvider/OracleConnectionAdapter.cs
using GrapeCity.BI.Data.DataProviders;
using Oracle.ManagedDataAccess.Client;
using System;
using System.Collections.Generic;
using System.Data.Common;
using System.Globalization;
using System.Linq;

namespace GrapeCity.ActiveReports.Samples.OracleDataProvider
{
	//implements additional features like setting of multivalue parameters and others
	public sealed class OracleConnectionAdapter : DbConnectionAdapter
	{
		public static OracleConnectionAdapter Instance = new OracleConnectionAdapter();

		/// <summary>
		/// Returns the string representation of a multi-value parameter's value.
		/// </summary>
		protected override string MultivalueParameterValueToString(object[] parameterArrayValue)
		{
			return string.Join(",", parameterArrayValue.Select(parameterValue => "'" + Convert.ToString(parameterValue, CultureInfo.InvariantCulture) + "'"));
		}

		/// <summary>
		/// Oracle Client command paremeters are always prefixed by : symbol, 
		/// </summary>
		/// <param name="name"></param>
		/// <returns></returns>
		protected override string CreateParameterNamePattern(string name)
		{
			if (!name.StartsWith(":"))
				name = string.Format(":{0}", name);
			return base.CreateParameterNamePattern(name);
		}

		/// <summary>
		/// Adds parameter values in to the command.
		/// </summary>
		/// <param name="queryParameters"></param>
		/// <param name="command"></param>
		public override void SetParameters(IEnumerable<DbCommandParameter> queryParameters, DbCommand command)
		{
			if (command is OracleCommand oracleCommand)
				oracleCommand.BindByName = true;

			base.SetParameters(queryParameters, command);
		}
	}
}
